### PR TITLE
SC-13 # `serverless` command translates project to Serverless structure

### PR DESCRIPTION
**Pull request by jokeyrhyme**
### Added
- SC-13: `bm server serverless --out <targetdir>` is a new command
  - undocumented, does not appear in `--help`
  - translates the project in the current working directory to the target output directory
  - uses a one-size-fits-all wrapper script and strict conventions to facilitate conversion
### Code Review Notes
- probably worth reviewing these one commit at a time
- known incomplete work to be tackled in future:
  - [ ] custom URLs
  - [ ] thorough test coverage (improved significantly in #6)
  - [ ] friendlier instructions and error messages
  - [ ] CORS
  - [ ] HTTP response headers
  - [ ] HTTP status codes / errors
  - [x] alternative `module.exports` conventions for HTTP methods (see #6)
  - [ ] native Node.js modules (compile for Linux)
  - [ ] project permissions managements
  - [ ] cloud-hosted proxy endpoint (with above permissions)

### CODE REVIEW

File: lib/serverless.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+function copyRecursive (source, target) {
+  return pify(cpr)(source, target, CPR_OPTIONS)
+}
+
+function copyWrapper (source, target) {
+  return apis.listAPIs(source)
+    .then((names) => Promise.all(names.map((name) => {
+      const targetWrapper = path.join(target, `${name}.js`)
+      return pify(cpr)(WRAPPER, targetWrapper, CPR_OPTIONS)
```
Comment (mymattcarroll) [inline lib/serverless.js]: Can this just be `return copyRecursive(WRAPPER, targetWrapper)`

File: lib/serverless.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    })))
+}
+
+function registerFunctions (source, target) {
+  const configPath = path.join(target, 'serverless.yml')
+  const config = yaml.safeLoad(fs.readFileSync(configPath), 'utf8')
```
Comment (mymattcarroll) [inline lib/serverless.js]: Is there a reason you're using Sync functions here and below?

**Review by mymattcarroll — COMMENTED** [review]
Just a couple of questions

File: lib/serverless.js (hunk 1 above)
Comment (jokeyrhyme) [reply lib/serverless.js]: Good call. I'll switch this over.

File: lib/serverless.js (hunk 2 above)
Comment (jokeyrhyme) [reply lib/serverless.js]: I got lazy. I'll switch these over to asynchronous / Promise stuff now.

**Comment by jokeyrhyme** [comment]
@mymattcarroll I've fixed up the DRY and sync issues: https://github.com/blinkmobile/server-cli/blob/SC-13/lib/serverless.js

File: lib/serverless.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+          })),
+          handler: `${name}.handler`
+        }
+        return functions
+      }, {})
+
+      return config
+    })
+    .then((config) => {
+      fs.writeFileSync(configPath, yaml.safeDump(config), 'utf8')
```
Comment (mymattcarroll) [inline lib/serverless.js]: I am assuming this was left as synchronous for reason?

Comment (jokeyrhyme) [reply lib/serverless.js]: Bah!

**Comment by jokeyrhyme** [comment]
@mymattcarroll okay, really really got them all this time :)

**Comment by mymattcarroll** [comment]
LGTM 👍

**Pull request merged**